addon/components/polaris-text-field.js

```diff
@@ -0,0 +1,520 @@
+   *
+   * @property onFocus
+   * @public
+   * @type {Function}
+   * @default noop
+   */
+  onFocus() {},
+
+  /**
+   * Callback when focus is removed
+   *
+   * @property onBlur
+   * @public
+   * @type {Function}
+   * @default noop
+   */
+  onBlur() {},
+
+  addValueListener() {
+    let [ field ] = this.element.querySelectorAll('input, textarea');
+    let _this = this;
+
+    field.addEventListener('keyup', () => {
+      debounceTask(_this, 'debouncedUpdateValue', 250);
+    });
+  },
+
+  removeValueListener() {
+    let [ field ] = this.element.querySelectorAll('input, textarea');
+
+    // inline method is passed so tests don't fail
+    let _this = this;
+    field.removeEventListener('keyup', () => {
+      debounceTask(_this, 'debouncedUpdateValue', 250);
+    });
+  },
+
+  debouncedUpdateValue() {
+    let [ field ] = this.element.querySelectorAll('input, textarea');
+    this.set('value', field.value);
+  },
+
+  textFieldClasses: computed('value', 'disabled', 'readOnly', 'error', 'multiline', 'focus', function() {
+    let { value, disabled, readOnly, error, multiline, focus } = this.getProperties('value', 'disabled', 'readOnly', 'error', 'multiline', 'focus');
+
+    let valueClass = value ? 'Polaris-TextField--hasValue' : '';
+    let disabledClass = disabled ? 'Polaris-TextField--disabled' : '';
+    let readOnlyClass = readOnly ? 'Polaris-TextField--readOnly' : '';
+    let errorClass = error ? 'Polaris-TextField--error' : '';
+    let multilineClass = multiline ? 'Polaris-TextField--multiline' : '';
+    let focusClass = focus ? 'Polaris-TextField--focus' : '';
+
+    return `Polaris-TextField ${ valueClass } ${ disabledClass } ${ readOnlyClass } ${ errorClass } ${ multilineClass } ${ focusClass }`;
+  }),
+
+  ariaDescribedBy: computed('error', 'helpText', 'id', function() {
+    let { error, helpText, id } = this.getProperties('error', 'helpText', 'id');
+    let describedBy = [];
+
+    if (error) {
+      describedBy.push(`${ id }Error`);
+    }
+
+    if (helpText) {
+      describedBy.push(`${ id }HelpText`);
+    }
+
+    return describedBy.join(' ');
+  }),
+
+  ariaLabelledBy: computed('id', function() {
+    return `${ this.get('id') }Label`;
+  }),
+
+  ariaInvalid: computed('error', function() {
+    return !!this.get('error');
+  }),
+
+  inputType: computed('type', function() {
+    let type = this.get('type');
+
+    return type === 'currency' ? 'text' : type;
+  }),
+
+  input: computed('multiline', function() {
+    let multiline = this.get('multiline');
+
+    return multiline ? 'textarea' : 'input';
+  }),
+
+  minimumLines: computed('multiline', function() {
+    let multiline = this.get('multiline');
+
+    return (typeof multiline === 'number') ? multiline : 1;
+  }),
+
+  heightStyle: computed('height', 'multiline', function() {
+    let { height, multiline } = this.getProperties('height', 'multiline');
+
+    return (height && multiline) ? htmlSafe(`height: ${ height }px`) : '';
```

Comment:
Is this better being `null` than an empty string?

Reply:
Works either way, I can update it.